### ISSUE TITLE
Understanding rebalancing of keys when new node is added or existing node is removed

### ISSUE THREAD

**Issue by ashishnegi** [chrislusf/vasto#15]
This is a question for understanding data movement during adding or removing a node to cluster.

In my understanding, `jump consistent hashing` does not hash `node` on the ring. It seems like their is no ring concept. The output of hash is a number in `[0, nodes)` range which maps to a node.

If a new node is added to cluster, how will we know which all keys need to move to this node ?
Will we have to go to `all` nodes, read `all` keys and see if they belong to this new node ?

Can you point me to code which does this ?

In `ring based consistent hashing`, new node will be in ring and will copy keys from next node to itself. So, it may require going over all keys on only `one` node. However, in real world, one node maps to many buckets on the ring, so we may have to go to multiple nodes in `ring based consistent hashing` as well.

**Comment by chrislusf** [comment]
> Will we have to go to all nodes, read all keys and see if they belong to this new node ?

Yes. (The read and filter happen on the source side. Only required data are transmitted.)

here is one place that has these logic: https://github.com/chrislusf/vasto/blob/master/cmd/store/store_grpc_server_bootstrap.go

**Comment by ashishnegi** [comment]
In this regard, with `ring consistent hashing` we can also store `ring-key` with the datastore. We can then index on `ring-key`. This will remove full db scan. This will make faster node join or data movement.
What are your thoughts on this ?

**Comment by chrislusf** [comment]
data repartitioning does not happen often. No need to optimize for it. With current implementation, the state of the cluster can be atomically switched. The simpler state can be easily reasoned about. It will be good to ensure latency.
